### DISCUSSION TITLE
ESP32-S3: how to enter bootloader without pressing BOOT button (machine.bootloader() doesn't work)?

### DISCUSSION THREAD

**Discussion by ned-pcs** [micropython/micropython#13402] [ESP32]
We're building devices that embed ESP32-S3 modules and haven't yet set up OTA updating. It's hard to get to the BOOT button on our board once we've built the devices. So we'd like to be able to enter the bootloader for firmware updates by issuing a micropython command. 

However, `machine.bootloader()` doesn't do this (it only enters a tight loop), and I can't get `esptool.py` to talk to the ESP32 after that. The USB descriptors look different if you use the BOOT/RESET buttons vs. `machine.bootloader()`:
  - this works (BOOT/RESET buttons): `/dev/ttyACM0 34:85:18:73:5F:54 303a:1001 Espressif USB JTAG/serial debug unit`
  - this doesn't (`machine.bootloader()`): `/dev/ttyACM0 123456 303a:4001 Espressif Systems Espressif Device`

Is there a reliable way to force bootloader entry from MicroPython without pushing buttons?

**Comment by robhamerling** [comment]
I'm sorry not being really helpful, but I have the same problem with a Seeed Studio **ESP32-S3.** 
Most ESP32 boards go into bootloader mode 'automagically' with esptool.py, but not this one. In that case 'mpremote bootloader' works always, but  not with this ESP32-S3. esptool.py always terminates with:
```
A fatal error occurred: Failed to connect to ESP32-S3: No serial data received.
```
So that leaves using the boot and reset buttons, but these are so tiny that it very difficult to push 'm and even more in the required sequence.
It looks to me that a correction in the behaviour of 'machine.bootloader()' is required (opening an issue?)

**Reply by robert-hh** [reply]
Same here with a LOLIN ESP32 S3.  Looking into the code, machine.bootloader() is just working for the ARDUINO_NANO_ESP32. For all other board the code just enters an endless loop.
For the LOLIN ESP32 S3 the esptool.py bootloader activation only works when using the UART interface, not when using the OTG interface.

**Comment by robhamerling** [comment]
I've tried programming using TX(pin 43) and RX(pin 44) of the XIAO ESP32-S3: no success.

**Reply by robert-hh** [reply]
Using UART, the board has to use the espressif 2 transistor logic to switch EN and GPIO0 for bootloader mode by switching DTR and RTS. Just attaching a USB/Serial bridge is not sufficient. The LOLIN ESP32 S3 has this logic attached to the UART interface.
But still there is a difference e.g. between the pure USB interface of a S3 and a C3 device. The C3 can be switched to bootloader mode by esptool.py, the S3 will not enter bootloader mode. I do not know if this is a problem of the S3 or of esptool.py.

**Comment by andrewleech** [comment]
The S3 has some teething issues with the jump in/out of bootloader functionality. 

I've implemented explicit support for the RTS/DTR bootloader jump in PR https://github.com/micropython/micropython/pull/15108, along with behind the scenes switching from tinyusb to USB serial JTAG mode to workaround one of the known firmware bugs with this. 


The automatic reboot to bootloader in esptool has a separate issue however that's addressed in the external pr linked there, however it got only partially merged into esptool so there's often still some glitches there. TLDR the first time you run esptool (either manually or via make deploy) it fails during the reboot to bootloader because the serial port drops out. Rerunning it then works to perform the flash.

On a related note, the current release of esptool can flash the S3 just fine, but it can't automatically reboot back to application after the flashing is finished. Currently you need to physically hit reset or power cycle the board. This has already been fixed in esptool master https://github.com/espressif/esptool/commit/0215786283660480e9ec85dd077e6fc2f46919e9 but isn't in a release yet; it'll be in 4.8 when that's out.

**Comment by Josverl** [comment]
@ned-pcs,

Is there a reason why you can't use the same method esptool uses ?
you can either so a subprocess.run or import esptool and do a deeper integration if you want to avoid loading/initializing the tool too frequently

I use the [subprocesrun approach in mpflash](https://github.com/Josverl/micropython-stubber/blob/main/src/mpflash/mpflash/flash/esp.py), and that works great across the different chipsets.

Edit: 
or does your implementation not include the auto-reset curcuitry ? in which case I understand you need a different option.
https://docs.espressif.com/projects/esptool/en/latest/esp32s3/advanced-topics/boot-mode-selection.html#automatic-bootloader

**Reply by ned-pcs** [reply]
We don't have a USB/serial interface chip to provide the extra signals.
